pyam/iiasa.py

```diff
@@ -45,51 +52,93 @@ def set_config(user, password, file=None):
-        if _creds is None:
-            logger.error(f"Could not read credentials from `{creds}`")
-        creds = _creds
-    else:
-        msg = (
-            "Passing credentials as clear-text is not allowed. "
-            "Please use `pyam.iiasa.set_config(<user>, <password>)` instead!"
-        )
-        raise DeprecationWarning(msg)
+class SceSeAuth(AuthBase):
+    def __init__(self, creds: str = None, auth_url: str = _AUTH_URL):
+        """Connection to the Scenario Services Manager AAC service.
+
+        Parameters
+        ----------
+        creds : pathlib.Path or str, optional
+            Path to a file with authentication credentials
+        auth_url : str, optionl
+            Url of the authentication service
+        """
+        self.client = httpx.Client(base_url=auth_url, timeout=10.0, http2=True)
+        self.access_token, self.refresh_token = None, None
+
+        if creds is None:
+            if DEFAULT_IIASA_CREDS.exists():
+                self.creds = _read_config(DEFAULT_IIASA_CREDS)
+            else:
+                self.creds = None
+        elif isinstance(creds, Path) or isstr(creds):
+            self.creds = _read_config(creds)
+        else:
+            raise DeprecationWarning(
+                "Passing credentials as clear-text is not allowed. "
+                "Please use `pyam.iiasa.set_config(<user>, <password>)` instead!"
+            )
+
+        # if no creds, get anonymous token
+        # TODO: explicit token for anonymous login will not be necessary for ixmp-server
+        if self.creds is None:
+            r = self.client.get("/legacy/anonym/")
+            if r.status_code >= 400:
+                raise ValueError("Unknown API error: " + r.text)
+            self.user = None
+            self.access_token = r.json()
 
-    # if (still) no creds, get anonymous auth and return
-    if creds is None:
-        url = "/".join([auth_url, "anonym"])
-        r = requests.get(url)
-        _check_response(r, "Could not get anonymous token")
-        return r.json(), None
+        # else get user-token
+        else:
+            self.user = self.creds["username"]
+            self.obtain_jwt()
+
+    def __call__(self):
+        try:
+            # raises jwt.ExpiredSignatureError if token is expired
+            jwt.decode(self.access_token, options=JWT_DECODE_ARGS)
+
+        except jwt.ExpiredSignatureError:
+            self.refresh_jwt()
 
-    # get user token
-    headers = {"Accept": "application/json", "Content-Type": "application/json"}
-    url = "/".join([auth_url, "login"])
-    r = requests.post(url, headers=headers, data=json.dumps(creds))
-    user = creds["username"]
-    _check_response(r, f"Login failed for user {user}")
-    return r.json(), user
+        return {"Authorization": "Bearer " + self.access_token}
+
+    def obtain_jwt(self):
+        r = self.client.post("/v1/token/obtain/", json=self.creds)
+        if r.status_code == 401:
+            raise ValueError(
```

Comment:
Maybe use self.auth_url here?

Reply:
If a (non-expert) user reads the error message and tries to put the auth-url into his/her browser, this returns an unhelpful "Not Found" page - so I would rather use a url that users can "use".